core/base/iterator_factory.hpp

```diff
@@ -34,355 +34,320 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+        --(*this);
+        return tmp;
+    }
+
+    zip_iterator operator+(difference_type i) const
+    {
+        auto tmp = *this;
+        tmp += i;
+        return tmp;
+    }
+
+    friend zip_iterator operator+(difference_type i, const zip_iterator& iter)
+    {
+        return iter + i;
+    }
+
+    zip_iterator operator-(difference_type i) const
+    {
+        auto tmp = *this;
+        tmp -= i;
+        return tmp;
+    }
+
+    difference_type operator-(const zip_iterator& other) const
+    {
+        return forall_check_consistent(
+            other, [](const auto& a, const auto& b) { return a - b; });
+    }
+
+    reference operator*() const
+    {
+        return deref_impl(std::index_sequence_for<Iterators...>{});
+    }
+
+    reference operator[](difference_type i) const { return *(*this + i); }
+
+    bool operator==(const zip_iterator& other) const
+    {
+        return forall_check_consistent(
+            other, [](const auto& a, const auto& b) { return a == b; });
+    }
+
+    bool operator!=(const zip_iterator& other) const
+    {
+        return !(*this == other);
+    }
+
+    bool operator<(const zip_iterator& other) const
+    {
+        return forall_check_consistent(
+            other, [](const auto& a, const auto& b) { return a < b; });
+    }
+
+    bool operator<=(const zip_iterator& other) const
+    {
+        return forall_check_consistent(
+            other, [](const auto& a, const auto& b) { return a <= b; });
+    }
+
+    bool operator>(const zip_iterator& other) const
+    {
+        return !(*this <= other);
+    }
+
+    bool operator>=(const zip_iterator& other) const
+    {
+        return !(*this < other);
     }
 
 private:
-    ToSortType* dominant_values_;
-    SecondaryType* secondary_values_;
-    size_type size_;
+    template <std::size_t... idxs>
+    reference deref_impl(std::index_sequence<idxs...>) const
+    {
+        return reference{std::get<idxs>(iterators_)...};
+    }
+
+    template <typename Functor>
+    void forall(Functor fn)
+    {
+        forall_impl(fn, index_sequence{});
+    }
+
+    template <typename Functor, std::size_t... idxs>
+    void forall_impl(Functor fn, std::index_sequence<idxs...>)
+    {
+        tuple_unpack_sink((fn(std::get<idxs>(iterators_)), 0)...);
+    }
+
+    template <typename Functor, std::size_t... idxs>
+    void forall_impl(const zip_iterator& other, Functor fn,
+                     std::index_sequence<idxs...>) const
+    {
+        tuple_unpack_sink(
+            (fn(std::get<idxs>(iterators_), std::get<idxs>(other.iterators_)),
+             0)...);
+    }
+
+    template <typename Functor>
+    auto forall_check_consistent(const zip_iterator& other, Functor fn) const
+    {
+        auto result =
+            fn(std::get<0>(iterators_), std::get<0>(other.iterators_));
+        forall_impl(
+            other, [&](auto a, auto b) { assert(fn(a, b) == result); },
+            index_sequence{});
+        return result;
+    }
```

Comment:
You can only compare two zip_iterators if all wrapped iterators return the same result for comparison and difference operators, otherwise you get inconsistent results that point to somebody mixing zip_iterators from two different sources.

If you want to see this in practice, take a look at `IncompatibleIteratorDeathTest`, where I'm testing this behavior explicitly

Reply:
How would the partial ordering work? It sounds to me like that would break the assumed properties of a zip iterator?

Reply:
Thanks for the feedback!